Update dictionary-expressions.md

CyrusNajmabadi opened the pull request.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -6,10 +6,10 @@
 1. The well-known generic BCL dictionary interface types: `IDictionary<TKey, TValue>` and `IReadOnlyDictionary<TKey, TValue>`.
 
-Further support is present for dictionary-like types not covered above through the `CollectionBuilderAttribute` and a similar API pattern to the corresponding *create method* pattern introduced for collection expressions.
+Further support is present for dictionary-like types not covered above through the `CollectionBuilderAttribute` and a similar API pattern to the corresponding *create method* pattern introduced for collection expressions.  For example `ImmutableDictionary<TKey, TValue>` and `FrozenDictionary<TKey, TValue>`.
```
```suggestion
Further support is present for dictionary-like types not covered above through the `CollectionBuilderAttribute` and a similar API pattern to the corresponding *create method* pattern introduced for collection expressions.  Types like `ImmutableDictionary<TKey, TValue>` and `FrozenDictionary<TKey, TValue>` will be updated to support this pattern.
```

cston commented on proposals/dictionary-expressions.md:
```diff
@@ -249,16 +147,25 @@ HashSet<string> values = [StringComparer.CaseInsensitive, .. names];
+// `comparer: ...` to indicate the purpose of this value
 Dictionary<string, int> caseInsensitiveMap = [comparer: StringComparer.CaseInsensitive, .. existingMap];
+
+// Semicolon to more clearly dilineate the comparer
```
Typo: "delineate"

cston commented on proposals/dictionary-expressions.md:
```diff
@@ -435,3 +464,5 @@ Working group recommendation: Use normal parsing here.  So this would be the sam
 
 It would likely make sense to align with whatever comes out if you `foreach`ed the collection.
+
+TODO: Working group recomendation: Look at what happens with existing *collection expressions* and *collection types* with similar cases.  Follow that.  
```
>Look at what happens with existing *collection expressions* and *collection types* with similar cases. [](http://example.com/codeflow?start=35&length=102)

A collection type with multiple implementations of `IEnumerable<T>` (and with no accessible `GetEnumerator()` method) cannot be used as a collection expression target type.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -249,25 +157,33 @@ HashSet<string> values = [StringComparer.CaseInsensitive, .. names];
+// `comparer: ...` to indicate the purpose of this value
 Dictionary<string, int> caseInsensitiveMap = [comparer: StringComparer.CaseInsensitive, .. existingMap];
+
+// Semicolon to more clearly dilineate the comparer
```
```suggestion
// Semicolon to more clearly delineate the comparer
```

cston commented on proposals/dictionary-expressions.md:
````diff
@@ -55,44 +55,54 @@ Choices here would have implications regarding potential syntactic ambiguities,
-```c#
-__result.Add(new KeyValuePair<string, int>("mads", 21));
-```
+2. Introduction of the *dictionary type*. *Dictionary types* are types that are similar to the existing *collection types*, with the additional requirements that they have an *element type* of some `KeyValuePair<TKey, TValue>` *and* have an indexer `TValue this[TKey] { ... }`. The former requirement ensures that `List<T>` is not considered a dictionary type, with its `int`-to-`T` indexer. The lat ...
````
For `List<T>`, if it's the former requirement, should the reason be the following?

> The former requirement ensures that `List<T>` is not considered a dictionary type, since the *element type* is `T`.

CyrusNajmabadi replied:
yup!

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
````diff
@@ -55,178 +55,86 @@ Choices here would have implications regarding potential syntactic ambiguities,
+    ```
 
-Yes.  These are legal: [LDM-2024-03-11](https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-03-11.md#conclusions)
+2. Introduction of the *dictionary type*. *Dictionary types* are types that are similar to the existing *collection types*, with the additional requirements that they have an *element type* of some `KeyValuePair<TKey, TValue>` *and* have an indexer `TValue this[TKey] { ... }`. The former requirement ensures that `List<T>` is not considered a dictionary type, with its `int`-to-`T` indexer. The lat ...
````
```suggestion
2. Introduction of the *dictionary type*. *Dictionary types* are types that are similar to the existing *collection types*, with the additional requirements that they have an *element type* of some `KeyValuePair<TKey, TValue>` *and* have an indexer `TValue this[TKey] { ... }`. The former requirement ensures that `List<T>` is not considered a dictionary type, as its element type is `T` not `KeyValuePair<,>`. The latter requirement ensures that `List<KeyValuePair<int, string>>` is also not considered a dictionary type, with its `int`-to-`KeyValuePair<int, string>` indexer (instead of an `int`-to-`string` indexer). `Dictionary<TKey, TValue>` passes both requirements.
```

CyrusNajmabadi commented:
@cston @MadsTorgersen @RikkiGibson @jnm2 this is ready for review (though i'll likely be merging soon no matter what).  I made the following changes:

1. fixed minor bugs.
2. added some clarifying text to certain areas.
3. pulled out questions we aren't even intending to ask for, and moved to a 'retracted' section.  
4. front loaded 'comparers' more heavily.
5. made certain decisions/explanations firmer.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -6,10 +6,10 @@
 
 Several dictionary-like types can created without external BCL support.  These types are:
```
whitespace off to help with the review.

cston commented on proposals/dictionary-expressions.md:
```diff
@@ -249,25 +157,33 @@ HashSet<string> values = [StringComparer.CaseInsensitive, .. names];
 * A *struct* or *class* *dictionary type* that implements `System.Collections.IEnumerable` where:
   * The *element type* is determined from a `GetEnumerator` instance method or enumerable interface.
-  * The *type* has an *[applicable](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#11642-applicable-function-member)* constructor that can be invoked with no arguments (*or* a constructor with a single parameter whose type is convertible to `IEqualityComparer<TKey>`), and the constructor is accessible at the location of the collection expression.
+  * The *type* has an *[applicable](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#11642-applicable-function-member)* constructor that can be invoked with no arguments (*or* a constructor with a single [*comparer*](Comparer-support) parameter), and the constructor is accessible at the location of the collection expression.
```
>(Comparer-support) [](http://example.com/codeflow?start=249&length=18)

If these are links to the "Comparer support" section of this document, the link should be <code>(#Comparer-support)</code>.

CyrusNajmabadi replied:
done.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -6,10 +6,10 @@
 
 Several dictionary-like types can created without external BCL support.  These types are:
 
-1. Concrete dictionary-like types, containing an indexer `TValue this[TKey] { get; }`, like `Dictionary<TKey, TValue>`, `ConcurrentDictionary<TKey, TValue>`, `ImmutableDictionary<TKey, TValue>`, and `FrozenDictionary<TKey, TValue>`.
```
i realized this was wrong.  ID and FD shouldn't be in this list.  They should be in the list we need runtime 'create method' support for.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -55,178 +55,86 @@ Choices here would have implications regarding potential syntactic ambiguities,
+There are three core aspects to the design of dictionary expressions. 
 
-First is the concept of a *dictionary type*. *Dictionary types* are types that are similar to the existing *collection types*, with the additional requirements that they have an *element type* of some `KeyValuePair<TKey, TValue>` *and* have an indexer `TValue this[TKey] { ... }`. The former requirement ensures that `List<T>` is not considered a dictionary type, with its `int`-to-`T` indexer. The  ...
+1. Collection expressions containing `KeyValuePair<,>` (coming from `expression_element`, `spread_element`, or `key_value_pair_element` elements) can now instantiate a normal *collection type* *or* a *dictionary type*.
```
moved to a form where we give an intuition, and then examples of it.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
````diff
@@ -55,178 +55,86 @@ Choices here would have implications regarding potential syntactic ambiguities,
+    ```
 
-### Answered question 4
+3. Alignment of the rules for assigning to *dictionary types* with the rules for assigning to *collection types*, just requiring aspects such as *element* and *iteration types* to be some `KeyValuePair<,>`.  *However*, with the rules extended such that that `KeyValuePair<,>` type itself is relatively transparent, and instead the rule is updated to work on the underlying `TKey` and `TValue` types.
````
added this new intuition.  Specifically that while `KVP<>` is the element type, we want to operate more on the level of Keys/Values.   I give an example below this of why that is desirable.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -55,178 +55,86 @@ Choices here would have implications regarding potential syntactic ambiguities,
-
-**Resolution:** *Spread elements* of key-value pair collections will be supported in dictionary expressions. [LDM-2024-03-11](https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-03-11.md#conclusions)
-
-### Answered question 3
```
Answered questions moved to the end.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -55,178 +55,86 @@ Choices here would have implications regarding potential syntactic ambiguities,
 * The *type* has an instance *indexer* where:
   * The indexer has a single parameter with an identity conversion from the parameter type to `TKey`.\*
-  * There is an identity conversion from the indexer type to `TValue`.\*
+  * There is an identity conversion from the indexer return type to `TValue`.\*
```
@cston  if this word is needed.  If the spec already talks about `indexer types` in other places, then we can revert.  I liked `return type` to make it clear which type this was referring to.

cston replied:
`TValue` is only the return type from the getter, so I'd suggest leaving as "type" rather than "return type".

CyrusNajmabadi replied:
done.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -249,25 +157,33 @@ HashSet<string> values = [StringComparer.CaseInsensitive, .. names];
 Dictionary<string, int> caseInsensitiveMap = [comparer: StringComparer.CaseInsensitive, .. existingMap];
+
+// Semicolon to more clearly delineate the comparer
+Dictionary<string, int> caseInsensitiveMap = [StringComparer.CaseInsensitive; .. existingMap];
```
based on suggestion from mads

cston commented on proposals/dictionary-expressions.md:
```diff
@@ -402,36 +316,152 @@ Synthesized types are free to employ any strategy they want to implement the req
+
+What are the rules when types have multiple indexers and multiple impls of `IEnumerable<KVP<,>>`?  
+
+This concern already exists with *collection types*.  For those types, the rule is that we must have an *element type* as per the existing language rules.  This follows for *dictionary types*, along wih the rule that there must be a corresponding indexer for this *element type*.  If those hold, the type can be used a a *dictionary type*.  If these don't hold, it cannot be.
```
```suggestion
This concern already exists with *collection types*.  For those types, the rule is that we must have an *element type* as per the existing language rules.  This follows for *dictionary types*, along with the rule that there must be a corresponding indexer for this *element type*.  If those hold, the type can be used as a *dictionary type*.  If these don't hold, it cannot be.
```

CyrusNajmabadi replied:
Also retracted.  This was solved just fine for normal collection types.  Nothing  needs to change here as far as we can tell.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -301,8 +217,8 @@ The new rules above represent a breaking change: For types that are a valid conv
->   - The method must have a single parameter of type System.ReadOnlySpan<E>, passed by value, and there is an identity conversion from E to the iteration type of the collection type.
->    - **The method has two parameters, where one is convertible to an `IEqualityComparer<TKey>` and the other follows the rules of the *single parameter* rule above. This method will be called if the collection expression's first element is an `expression_element` that is convertible to that parameter type.**
+>   - The method must have a single parameter of type System.ReadOnlySpan<E>, passed by value, and there is an identity conversion from E to the *iteration type* of the collection type. 
+>    - **The method can be called with two parameters, where one is a [*comparer*](Comparer-support) and the other follows the rules of the *single parameter* rule above. This method will be called if the collection expression's first element is an [*comparer*](Comparer-support) that is convertible to that parameter type.**
```
note: i changed the wording subtly.  For constructors we just say it needs to be callable (meaning it could have optional arguments).  We can return this to just be exactly two parameters.  We can also be explicit if the comparer has to be the first parameter or not.  

This part is TBD when talking to runtime.  It will come down to if they want exactly one pattern used *everywhere*, or if they have a couple they expect.

Note: ToFrozenDictionary is of the form:

```c#
public static FrozenDictionary<TKey, TValue> ToFrozenDictionary<TKey, TValue>(this IEnumerable<KeyValuePair<TKey, TValue>> source, IEqualityComparer<TKey>? comparer = null)
```

While immutable dictionary looks like:

```
public static ImmutableDictionary<TKey, TValue> CreateRange<TKey, TValue>(IEqualityComparer<TKey>? keyComparer, IEnumerable<KeyValuePair<TKey, TValue>> items)
```

So, if we want to point at existing methods, we'd need to support these forms.  However, if we're mandating `ROS<KVP>`, then we'll need new overloads anyways, and we can have strict rules on the shape.

--

Finally, we may likelyy need new overloads as FrozenDict has overwrite semantics (which we want) while ImmutableDict does not (it throws).  

Tagging in @stephentoub .

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -312,11 +228,11 @@ The runtime has committed to supplying these new CollectionBuilder methods that
-3. If creating an interface, this comparer will be used as the equality comparer controlling the behavior of the final type (synthesized or otherwise) instantiated.
+1. If using a constructor to instantiate the value, the constructor must take a single parameter whose type is some [*comparer*](Comparer-support) type.  The first `element_expression` value will be passed to this parameter.
+2. If using a `create method`, the method must have a parameter whose type is some [*comparer*](Comparer-support) as one of its parameters. The first `element_expression` value will be passed to this parameter.
+3. If creating an interface, this [*comparer*](Comparer-support) must be some `IEqualityComparer<TKey>` type. That comparer will be used to  control the behavior of the final type (synthesized or otherwise).  This means that instantiating interfaces only supports hashing semantics, not ordered semantics. 
```
Tagging @stephentoub .  If you guys added an `ISortedDictionary<>` and `IReadOnlySortedDictionary<>` we could key off of those to create sorted versions automatically.  That can also come in the future.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -402,36 +316,152 @@ Synthesized types are free to employ any strategy they want to implement the req
+## Retracted Designs/Questions
 
-There is a subtle concern around the following interface destinations:
+### Question: Should `k:v` elements force dictionary semantics?
```
No one is coming with this recommendation anymore.  And we made a firm LDM stance in the last meeting that `k:v` was fine for normal non-dicttionary types.  So this question is retracted.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -402,36 +316,152 @@ Synthesized types are free to employ any strategy they want to implement the req
+### Question: Allow deconstructible types?
 
-Parsing ambiguity around: `[a ? [b] : c]`
+Should we take a very restrictive view of `KeyValuePair<,>`?  Specifically, should we allow only that exact type?  Or should we allow any types with an implicit conversion to that type?  For example:
```
This idea made a lot of people uncomfortable.  And we do not need it at all for a viable and good dictionary story.  So explicitly retracting this from the design.  This can always be revisited later.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -402,36 +316,152 @@ Synthesized types are free to employ any strategy they want to implement the req
+2. Use applicable instance indexer if available; otherwise report an error during construction (or conversion?).
+3. Use C#12 initialization always.
+
+Resolution TBD.  Working group recommendation: Use applicable instance indexer only.  This ensures that everything dictionary-like is initialized in a consistent fashion.  This would be a break in behavior when recompiling.  The view is that these types would be rare.  And if they exist, it would be nonsensical for them to behave differently using the indexer versus the .Add (outside of potential ...
```
this is still a good open question, after 'comparers'.  We should get LDM thoughts on how we want to proceed here.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -402,36 +316,152 @@ Synthesized types are free to employ any strategy they want to implement the req
+
+Parsing ambiguity around: `[a ? [b] : c]`
+
+Working group recommendation: Use normal parsing here.  So this would be the same as `[a ? ([b]) : (c)]` (a collection expression containing a conditional expression).  If the user wants a `key_value_pair_element` here, they can write: `[(a?[b]) : c]`
```
very low level concern.  Just want to get sign off that the direction the existing impl takes is the desired one.

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -55,178 +55,86 @@ Choices here would have implications regarding potential syntactic ambiguities,
 * The *type* has an instance *indexer* where:
   * The indexer has a single parameter with an identity conversion from the parameter type to `TKey`.\*
-  * There is an identity conversion from the indexer type to `TValue`.\*
+  * There is an identity conversion from the indexer return type to `TValue`.\*
```
```suggestion
  * There is an identity conversion from the indexer type to `TValue`.\*
```

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -312,11 +228,11 @@ The runtime has committed to supplying these new CollectionBuilder methods that
-2. If using a `create method`, the method must have a parameter whose type is convertible to `IEqualityComparer<TKey>` as one of its parameters. The first `element_expression` value will be passed to this parameter.
-3. If creating an interface, this comparer will be used as the equality comparer controlling the behavior of the final type (synthesized or otherwise) instantiated.
+1. If using a constructor to instantiate the value, the constructor must take a single parameter whose type is some [*comparer*](#Comparer-support) type.  The first `element_expression` value will be passed to this parameter.
+2. If using a `create method`, the method must have a parameter whose type is some [*comparer*](#Comparer-support) as one of its parameters. The first `element_expression` value will be passed to this parameter.
```
```suggestion
2. If using a `create method`, the method's first parameter's type is some [*comparer*](#Comparer-support) type. The first `element_expression` value will be passed to this parameter.
```

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -301,8 +217,8 @@ The new rules above represent a breaking change: For types that are a valid conv
->   - The method must have a single parameter of type System.ReadOnlySpan<E>, passed by value, and there is an identity conversion from E to the iteration type of the collection type.
->    - **The method has two parameters, where one is convertible to an `IEqualityComparer<TKey>` and the other follows the rules of the *single parameter* rule above. This method will be called if the collection expression's first element is an `expression_element` that is convertible to that parameter type.**
+>   - The method must have a single parameter of type System.ReadOnlySpan<E>, passed by value, and there is an identity conversion from E to the *iteration type* of the collection type. 
+>    - **The method can be called with two parameters, where one is a [*comparer*](#Comparer-support) and the other follows the rules of the *single parameter* rule above. This method will be called if the collection expression's first element is an [*comparer*](#Comparer-support) that is convertible to that parameter type.**
```
```suggestion
>    - **The method has two parameters, where the first is a [*comparer*](#Comparer-support) and the other follows the rules of the *single parameter* rule above. This method will be called if the collection expression's first element is an [*comparer*](#Comparer-support) that is convertible to that parameter type.**
```

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -301,8 +217,8 @@ The new rules above represent a breaking change: For types that are a valid conv
 > **A create method will commonly use the name `CreateRange` in the dictionary domain.**
```
```suggestion
> **A create method need not be called `Create`.  Instead, it may commonly use the name `CreateRange` in the dictionary domain.**
```

CyrusNajmabadi commented on proposals/dictionary-expressions.md:
```diff
@@ -301,8 +217,8 @@ The new rules above represent a breaking change: For types that are a valid conv
+>   - The method must have a single parameter of type System.ReadOnlySpan<E>, passed by value, and there is an identity conversion from E to the *iteration type* of the collection type. 
+>    - **The method has two parameters, where the first is a [*comparer*](#Comparer-support) and the other follows the rules of the *single parameter* rule above. This method will be called if the collection expression's first element is an [*comparer*](#Comparer-support) that is convertible to that parameter type.**
 
 This would allow `ImmutableDictionary<TKey, TValue>` to be annotated with `[CollectionBuilder(typeof(ImmutableDictionary), "CreateRange")]` to light up support for creation.
```
```suggestion
Dictionary type authors who use `CollectionBuilderAttribute` should have the method that is pointed to have `overwrite` not `throw` semantics when encountering the same `.Key` multiple times in the span of `KeyValuePair<,>` they are processing.
```
